stackrox/stackrox
# feat(cloud-sources): add OCM support

## Description

This introduces OCM support for cloud sources, and covers the following pieces:
- Adding the cluster ID to the cluster metadata collected by Sensor, this is required for matching discovered clusters with secured clusters (where the type is OCP, ARO, ROSA, OSD).
- implement the client for interacting with the OCM API.
- improve the matching to only rely on the ID for matching instead of the tuple of ID & provider type (we can match on ID only since for our known cluster types the IDs have a distinct format irrespective of cluster type).

A few caveats to point out:
The `accounts_mgmt` API was used in favor of the `clusters_mgmt` API; I have found that this is way more performant compared to the cluster one, and also what `console.redhat.com/openshift` does.

The search query to limit the result of clusters (or rather subscriptions) has been taken from `console.redhat.com/openshift` with an adjustment to the `Status` field: we currently only take `Active, Disconnected` clusters into account. We may extend this to also cover other states (such as `Installing`); or we might also want to lower the number of allowed states to only `Active` and `Disconnected`.

The OCM connection currently does not take in the `proxy.RoundTripper()` directly, but rather relies on the call of `proxy.UseWithDefaultTransport` within the main to set the proxy transport that way. The OCM SDK currently only provides a `RoundTripWrapper`. The alternative would be to initialize the accounts client ourselves, but then we lose the benefit of the authN and other wrappers the SDK provides us with.

What still needs to be done after this has landed:
The UI needs to also support the new integration type by providing a new integration form.

## Checklist

If any of these don't apply, please comment below.

## Testing Performed

### Here I tell how I validated my change

- see CI.

Manual tests:

1. Create a Central cluster + secured cluster (use a provider of choice, e.g. ROSA/ARO)

2. Create an OCM cloud source; use an Offline API token for OCM:
```json
{
    "cloudSource": {
        "name": "test ocm",
        "type": "TYPE_OCM",
        "credentials": {
            "secret": "<redacted>"
        },
        "ocm": {
            "endpoint": "https://api.openshift.com"
        }
    }
}
```

3. Once the discovered clusters have synced, you should see that the discovered cluster is shown as secured.


### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.
